src/backend/core/models.py

```diff
@@ -155,6 +176,195 @@ def get_teams(self):
+        ordering = ("title",)
+        verbose_name = _("Document")
+        verbose_name_plural = _("Documents")
+
+    def __str__(self):
+        return self.title
+
+    @property
+    def content(self):
+        """Return the json content from object storage if available"""
+        if self._content is None and self.id:
+            try:
+                # Load content from object storage
+                with default_storage.open(str(self.id)) as f:
+                    self._content = json.load(f)
+            except FileNotFoundError:
+                pass
+        return self._content
+
+    @content.setter
+    def content(self, content):
+        """Cache the content, don't write to object storage yet"""
+        if isinstance(content, str):
+            content = json.loads(content)
+        if not isinstance(content, dict):
+            raise ValueError("content should be a json object.")
+        self._content = content
+
+    def save(self, *args, **kwargs):
+        """Write content to object storage only if _content has changed."""
+        super().save(*args, **kwargs)
+
+        if self._content:
+            file_key = str(self.pk)
+            bytes_content = json.dumps(self._content).encode("utf-8")
+
+            if default_storage.exists(file_key):
+                response = default_storage.connection.meta.client.head_object(
+                    Bucket=default_storage.bucket_name, Key=file_key
+                )
+                has_changed = (
+                    response["ETag"].strip('"')
+                    != hashlib.md5(bytes_content).hexdigest()  # noqa
+                )
+            else:
+                has_changed = True
+            if has_changed:
+                content_file = ContentFile(bytes_content)
+                default_storage.save(file_key, content_file)
+
+    def get_abilities(self, user):
+        """
+        Compute and return abilities for a given user on the document.
+        """
+        roles = get_resource_roles(self, user)
+        is_owner_or_admin = bool(
+            set(roles).intersection({RoleChoices.OWNER, RoleChoices.ADMIN})
+        )
+        can_get = self.is_public or bool(roles)
+
+        return {
+            "destroy": RoleChoices.OWNER in roles,
+            "manage_accesses": is_owner_or_admin,
+            "update": is_owner_or_admin,
+            "retrieve": can_get,
+        }
+
+
+class DocumentAccess(BaseAccess):
+    """Relation model to give access to a document for a user or a team with a role."""
+
+    resource = models.ForeignKey(
+        Document,
+        on_delete=models.CASCADE,
+        related_name="accesses",
+    )
+
+    class Meta:
+        db_table = "impress_document_access"
+        ordering = ("-created_at",)
+        verbose_name = _("Document/user relation")
+        verbose_name_plural = _("Document/user relations")
+        constraints = [
+            models.UniqueConstraint(
+                fields=["user", "resource"],
```

Comment:
Is it possible to have `document_id` instead of `resource_id` ?

Reply:
yes... it will make heritage a little less direct but if you think it is better I can revert to it!

Reply:
I find it is better to have the table indicated in the foreign key , after depend if it is complicated. Can you do polymorphic inheritance with drf ?
I'm fine with `resource_id` as well, don't spent too much time on that.

Reply:
done

Reply:
Nice